### PR TITLE
Report in LIMS when a pipeline starts or finishes in a sample

### PR DESCRIPTION
**Pull request by katieemelianova**
Fix run_qc bug
update egcg_core to 0.7 in requirements

### CODE REVIEW

File: analysis_driver/dataset.py [hunk 1]
```diff
@@ -65,19 +72,23 @@ def start(self):
...
 
     def fail(self, exit_status):
         self._assert_status(DATASET_PROCESSING)
         self.ntf.end_pipeline(exit_status)
+        if self.type == 'sample': self.lims_ntf.fail_and_reset_sample_pipeline()
         self.most_recent_proc.finish(DATASET_PROCESSED_FAIL)
 
     def abort(self):
         self.most_recent_proc.finish(DATASET_ABORTED)
+        if self.type == 'sample': self.lims_ntf.fail_and_reset_sample_pipeline()
```
Comment (mwhamgenomics) [inline analysis_driver/dataset.py]: I'd have these in SampleDataset as additions to the default behaviour:

```python
class SampleDataset(Dataset):
    def abort(self):
        super().abort()
        self.lims_ntf.fail_and_reset_sample_pipeline()
```

File: analysis_driver/notification.py [hunk 2]
```diff
@@ -1,4 +1,59 @@
-from egcg_core import notifications
+from egcg_core import notifications, clarity
+from pyclarity_lims.lims import Step
```
Comment (mwhamgenomics) [inline analysis_driver/notification.py]: This should be pyclarity_lims.entities. Also, this should be added to requirements.txt

**Review by mwhamgenomics — CHANGES_REQUESTED** [review]

File: analysis_driver/dataset.py [hunk 3]
```diff
@@ -444,6 +440,22 @@ def report(self):
...
+        super().start()
+        self.lims_ntf.start_sample_pipeline()
+
+    def abort(self):
+        super().abort()
+        self.lims_ntf.fail_and_reset_sample_pipeline()
+
+    def succeed(self):
+        super().succeed()
+        self.lims_ntf.fail_and_reset_sample_pipeline()
```
Comment (mwhamgenomics) [inline analysis_driver/dataset.py]: Failing and resetting a successful pipeline?

**Review by mwhamgenomics — CHANGES_REQUESTED** [review]

File: analysis_driver/dataset.py (hunk 3 above)
Comment (katieemelianova) [reply analysis_driver/dataset.py]: Oops! Thanks for pointing that out, have changed it now

File: tests/test_dataset.py [hunk 4]
```diff
@@ -348,6 +352,41 @@ def setup_dataset(self):
         ]
         self.dataset._data_threshold = 1000000000
 
+    @patched_initialise
+    @patch(ppath + 'MostRecentProc.start')
+    @patch(ppath + 'MostRecentProc.retrieve_entity')
+    @patch(ppath + 'Dataset.lims_ntf', new_callable=PropertyMock)
+    def test_start(self, mocked_lims_ntf, mocked_retrieve_entity, mocked_start, mocked_update):
+        self.dataset.most_recent_proc.entity['status'] = c.DATASET_PROCESSING
```
Comment (mwhamgenomics) [inline tests/test_dataset.py]: Don't see why we're setting an item here when we're just `del`-ing it two lines later

**Review by mwhamgenomics — CHANGES_REQUESTED** [review]

File: tests/test_dataset.py (hunk 4 above)
Comment (katieemelianova) [reply tests/test_dataset.py]: I forgot to delete that line, thanks for pointing out

**Review by mwhamgenomics — APPROVED** [review]

File: analysis_driver/notification.py [hunk 5]
```diff
@@ -1,4 +1,59 @@
-from egcg_core import notifications
+from egcg_core import notifications, clarity
+from pyclarity_lims.entities import Step
+from analysis_driver.exceptions import PipelineError
+
+
+class LimsNotification():
+    def __init__(self, name):
+        self.name = name
+        self.sample_list = clarity.get_list_of_samples(self.name)
```
Comment (tcezard) [inline analysis_driver/notification.py]: We should only retrieve one and only one sample here.
`clarity.get_sample()` would be more appropriate and we should check the return is not `None`

File: analysis_driver/notification.py [hunk 6]
```diff
@@ -1,4 +1,59 @@
...
+        self.artifacts = [sample.artifact for sample in self.sample_list]
```
Comment (tcezard) [inline analysis_driver/notification.py]: Same as above, only one `artifact`

File: analysis_driver/notification.py [hunk 7]
```diff
@@ -1,4 +1,59 @@
...
+        self.step = None
+
+    @property
+    def lims(self):
+        return clarity.connection()
+
+    @property
+    def postseqlab_stages(self):
+        return self.lims.get_workflows(name='PostSeqLab EG 1.0 WF')[0].stages
```
Comment (tcezard) [inline analysis_driver/notification.py]: Put the workflow name in the config file.

File: analysis_driver/notification.py [hunk 8]
```diff
@@ -1,4 +1,59 @@
...
+        stage = [stage for stage in self.postseqlab_stages if stage == stage_name]
+        if not len(stage) == 1:
+            raise PipelineError('Can not find %s from Workflow Stages' % stage_name)
+        return stage[0]
+
+    def route_sample_to_data_processing(self):
+        self.lims.route_artifacts(self.artifacts, stage_uri=self.get_stage('Data Processing EG 1.0 ST').uri)
+
+    def create_step(self):
+        self.step = Step.create(self.lims, protocol_step=self.get_stage('Data Processing EG 1.0 ST').step, inputs=[self.artifacts], container_type_name='Tube')
```
Comment (tcezard) [inline analysis_driver/notification.py]: If `self.artifacts` is a list then you would pass a list of of one list to inputs.
Make sure the artifact is properly passed to `route_artifacts` and  `inputs`

File: analysis_driver/notification.py [hunk 9]
```diff
@@ -1,4 +1,59 @@
...
+
+    @property
+    def lims(self):
+        return clarity.connection()
+
+    @property
+    def postseqlab_stages(self):
+        return self.lims.get_workflows(name='PostSeqLab EG 1.0 WF')[0].stages
+
+    def get_stage(self, stage_name):
```
Comment (tcezard) [inline analysis_driver/notification.py]: The stage is the same across the entire run and does not need to be searched for every time.

File: analysis_driver/dataset.py [hunk 10]
```diff
@@ -31,6 +29,7 @@ def __init__(self, name, most_recent_proc=None):
         self.name = name
         self.most_recent_proc = MostRecentProc(self.type, self.name, most_recent_proc)
         self._ntf = None
+        self._lims_ntf = None
```
Comment (tcezard) [inline analysis_driver/dataset.py]: Does this need to be in `Dataset` if it is only going to be used in the `SampleDataset`

**Review by tcezard — CHANGES_REQUESTED** [review]

**Review by tcezard — CHANGES_REQUESTED** [review]
Also LimsNotification is not tested.
Can you add unit tests

File: tests/test_dataset.py [hunk 11]
```diff
@@ -348,6 +352,32 @@ def setup_dataset(self):
...
         self.dataset._data_threshold = 1000000000
 
+    @patched_initialise
+    @patch(ppath + 'MostRecentProc.start')
+    @patch(ppath + 'MostRecentProc.retrieve_entity')
+    @patch(ppath + 'SampleDataset.lims_ntf', new_callable=PropertyMock)
+    def test_start(self, mocked_lims_ntf, mocked_retrieve_entity, mocked_start, mocked_update):
+        with patched_get_run:
+            self.dataset.start()
+        self.assertTrue(mocked_lims_ntf.called)
```
Comment (tcezard) [inline tests/test_dataset.py]: This doesn't test that `start_sample_pipeline` was called

File: tests/test_dataset.py [hunk 12]
```diff
@@ -348,6 +352,32 @@ def setup_dataset(self):
...
+
+    @patched_finish
+    @patch(ppath + 'MostRecentProc.retrieve_entity')
+    @patch(ppath + 'SampleDataset.lims_ntf', new_callable=PropertyMock)
+    def test_succeed(self, mocked_lims_ntf, mocked_retrieve_entity, mocked_finish):
+        with patched_get_run:
+            self.dataset.most_recent_proc.entity['status'] = c.DATASET_PROCESSING
+            self.dataset.succeed()
+        self.assertTrue(mocked_lims_ntf.called)
```
Comment (tcezard) [inline tests/test_dataset.py]: This doesn't test that `assign_next_and_advance_step` was called

**Review by tcezard — CHANGES_REQUESTED** [review]

File: tests/test_dataset.py [hunk 13]
```diff
@@ -336,24 +320,6 @@ def test_no_data_threshold(self, mocked_exp_yield):
...
-        exp = {
-            'name': 'qc',
-            'toolset_type': 'non_human_sample_processing',
-            'toolset_version': 0
-        }
-
-        with patched_get_doc(None), patched_patch as mocked_patch:
-            assert self.dataset._pipeline_instruction() == exp
-            mocked_patch.assert_called_with('projects', {'sample_pipeline': exp}, 'project_id', 'a_project')
-
```
Comment (mwhamgenomics) [inline tests/test_dataset.py:353]: Hey, we needed that!

File: tests/test_dataset.py [hunk 14]
```diff
@@ -179,25 +174,14 @@ def test_raise_exceptions(self):
...
-            'toolset_version': 0
-        }
+
+class _TestDataset(Dataset):
+    type = 'None'
+    endpoint = 'None'
+    id_field = 'None'
+
+    def _is_ready(self):
+        pass
```
Comment (mwhamgenomics) [inline tests/test_dataset.py:184]: This was removed in master, but seems to have crept back in during the merge. Can this be reverted, along with the removal of `patched_is_ready`?

File: tests/test_dataset.py [hunk 15]
```diff
@@ -179,25 +174,14 @@ def test_raise_exceptions(self):
...
-        assert self.dataset.pipeline.name == 'qc'
-
-    @patch(ppath + 'toolset', new=Toolset())
-    def test_pipeline_instruction(self):
-        self.dataset._default_pipeline = Mock(return_value='qc')
-        assert self.dataset._pipeline_instruction() == {
-            'name': 'qc',
-            'toolset_type': 'non_human_sample_processing',
-            'toolset_version': 0
-        }
```
Comment (mwhamgenomics) [inline tests/test_dataset.py:197]: Hey, we needed that!

**Review by mwhamgenomics — CHANGES_REQUESTED** [review]
Some changes from master were overwritten. These should be restored.

**Pull request closed without merging**